src/datafusion/helpers.rs

```diff
@@ -243,3 +271,337 @@ pub(crate) fn build_row_filter(
+        let pf = PartitionedFile::new(table_path.prefix().clone(), 0);
+        return Ok(Box::pin(stream::iter(vec![Ok(pf)])));
+    }
+
+    let partitions = list_partitions(store, table_path, partition_cols.len()).await?;
+    let pruned = prune_partitions(table_path, partitions, filters, partition_cols).await?;
+
+    let stream = futures::stream::iter(pruned)
+        .map(move |partition: Partition| async move {
+            let cols = partition_cols.iter().map(|x| x.0.as_str()).collect();
+            let parsed = parse_partitions_for_path(table_path, &partition.path, cols);
+
+            let partition_values = parsed
+                .into_iter()
+                .flatten()
+                .zip(partition_cols)
+                .map(|(parsed, (_, datatype))| {
+                    ScalarValue::try_from_string(parsed.to_string(), datatype)
+                })
+                .collect::<DataFusionResult<Vec<_>>>()?;
+
+            let mut pf = PartitionedFile::new(partition.path, 0);
+            pf.partition_values.clone_from(&partition_values);
+
+            Ok(pf)
+        })
+        .buffer_unordered(CONCURRENCY_LIMIT)
+        .boxed();
+
+    Ok(stream)
+}
+
+// copied from datafusion
+pub fn split_files(
+    mut partitioned_files: Vec<PartitionedFile>,
+    n: usize,
+) -> Vec<Vec<PartitionedFile>> {
+    if partitioned_files.is_empty() {
+        return vec![];
+    }
+
+    // ObjectStore::list does not guarantee any consistent order and for some
+    // implementations such as LocalFileSystem, it may be inconsistent. Thus
+    // Sort files by path to ensure consistent plans when run more than once.
+    partitioned_files.sort_by(|a, b| a.path().cmp(b.path()));
+
+    // effectively this is div with rounding up instead of truncating
+    let chunk_size = (partitioned_files.len() + n - 1) / n;
+    partitioned_files
+        .chunks(chunk_size)
+        .map(|c| c.to_vec())
+        .collect()
+}
+
+#[cfg(test)]
+mod helpers_tests {
+    use super::*;
+    use crate::tests::get_test_v2_data_path;
+    use datafusion_expr::{and, col, lit};
+    use itertools::Itertools;
+    use object_store::local::LocalFileSystem;
+
+    #[tokio::test]
+    async fn test_listing_and_pruning_partitions() {
+        let table_path = get_test_v2_data_path("lat_lon_w_groups_example.zarr".to_string())
+            .to_str()
+            .unwrap()
+            .to_string();
+
+        let store = LocalFileSystem::new();
+        let url = ListingTableUrl::parse(table_path).unwrap();
+        let partitions = list_partitions(&store, &url, 2).await.unwrap();
+
+        let expr1 = col("var").eq(lit(1_i32));
+        let expr2 = col("other_var").eq(lit::<String>("b".to_string()));
+        let partition_cols = [
+            ("var".to_string(), DataType::Int32),
+            ("other_var".to_string(), DataType::Utf8),
+        ];
+
+        let prefix = "home/max/Documents/repos/arrow-zarr/test-data/data/zarr/v2_data/lat_lon_w_groups_example.zarr";
```

Comment:
Would be nice to make this relative to the package.

Reply:
Oh, yeah I didn't mean to do that, good catch. I got sloppy and forgot about that line I probably added for some intermediate test I ran.